sql: enforce failure on NULLs in RLS WITH CHECK expressions

spilchen opened the pull request:
Previously, row-level security (RLS) WITH CHECK constraints treated NULL values similarly to standard CHECK constraints, allowing NULL results to pass. This behavior was inconsistent with postgres and introduced potential security loopholes.

This change ensures that NULL results cause the constraint to fail.

Closes #143125

Epic: CRDB-45203
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143133)
<!-- Reviewable:end -->

spilchen committed:
sql: enforce failure on NULLs in RLS WITH CHECK expressions

Previously, row-level security (RLS) WITH CHECK constraints treated NULL
values similarly to standard CHECK constraints, allowing NULL results to
pass. This behavior was inconsistent with postgres and introduced
potential security loopholes.

This change ensures that NULL results cause the constraint to fail.

Closes #143125

Epic: CRDB-45203
Release note: none

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143133#-:-OLjBrB9B7I2dRyB5guu:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rytaft)



<!-- Sent from Reviewable.io -->
